### PR TITLE
fix(security): apply per-endpoint rate limiters to GraphQL auth mutations (#315)

### LINKED ISSUE #315: [Security] high: GraphQL endpoint bypasses all per-endpoint rate limiters (sign-in, refresh, password-reset, 2FA)
## Summary

GraphQL endpoint bypasses all per-endpoint rate limiters (sign-in, refresh, password-reset, 2FA)

**Severity:** HIGH  •  **CWE:** CWE-799  •  Found by automated multi-agent security audit (adversarially verified).

## Affected code

- `config/packages/security.yaml`
- `src/Shared/Application/Resolver/RateLimit/ApiRateLimitAuthTargetResolver.php`
- `src/Shared/Application/Resolver/RateLimit/ApiRateLimitRequestResolver.php`
- `src/User/Application/Resolver/ConfirmPasswordResetMutationResolver.php`
- `src/User/Application/Resolver/RefreshTokenAuthMutationResolver.php`
- `src/User/Application/Resolver/SignInAuthMutationResolver.php`

## Finding 1: GraphQL endpoint bypasses all per-endpoint rate limiters (sign-in, refresh, password-reset, 2FA)

**Class:** Improper Rate Limiting  •  **CWE:** CWE-799  •  **Verifier votes:** 2/2  •  **Confidence:** 0.8

**Locations:** `src/Shared/Application/Resolver/RateLimit/ApiRateLimitRequestResolver.php:47`, `src/Shared/Application/Resolver/RateLimit/ApiRateLimitAuthTargetResolver.php:48`, `src/User/Application/Resolver/SignInAuthMutationResolver.php:26`, `src/User/Application/Resolver/RefreshTokenAuthMutationResolver.php:31`, `src/User/Application/Resolver/ConfirmPasswordResetMutationResolver.php`, `config/packages/security.yaml:16`

### Description
All endpoint-specific rate limiters are selected by matching the REST URL path (e.g. resolveSignInLimiters checks path === '/api/signin', resolveTokenExchangeLimiter checks '/api/token', resolvePasswordResetConfirmLimiter checks '/api/reset-password/confirm', resolveSignInTwoFactorLimiters checks '/api/signin/2fa'). API Platform also exposes these same operations as GraphQL mutations (signIn, refreshToken, completeTwoFactor, confirmPasswordReset, signOut, signOutAll — see config/api_platform/resources/AuthPayload.yaml and User.yaml), all served at POST /api/graphql, which is PUBLIC_ACCESS (security.yaml line 16). For /api/graphql ApiRateLimitRequestResolver::resolveEndpointLimiters returns no endpoint limiters; only resolveGlobalLimiter (global_api_anonymous, default 100 req/min per IP) applies. Therefore the tight controls — signin_ip (10/min), signin_email (5/min), refresh_token (per IP), password_reset_confirm, twofa_verification_* — are all skipped when the same actions are performed through GraphQL.

### Exploit scenario
An attacker performs credential-stuffing / password-spraying against the `signIn` GraphQL mutation instead of REST /api/signin. Per-email throttling (5/min) and per-IP signin throttling (10/min) never fire; the attacker gets up to ~100 attempts/min per IP (per-account ultimately bounded only by the 20-failures/hour email lockout, which is much looser than 5/min). The same bypass applies to confirmPasswordReset token guessing and refreshToken abuse, which lose their dedicated limiters entirely. Combined with the 2FA finding, completeTwoFactor becomes brute-forceable.

### Suggested fix direction
Make rate-limit resolution operate on the logical operation, not the REST path. Parse the GraphQL query/operation name in the rate-limit listener and map signIn/refreshToken/completeTwoFactor/confirmPasswordReset/etc. to the same limiter keys, or move enforcement into the command/mutation layer so REST and GraphQL share one throttle. At minimum apply a strict per-IP and per-identity limiter to /api/graphql for these sensitive mutations.

---
_Reported by automated pentest audit. To be remediated via BMAD method (bmalph) — one PR per issue, CI green + AI review approval required._

### PR DESCRIPTION
**Pull request by dmytrocraft**
### Vulnerability (HIGH, CWE-799)
Per-endpoint authentication rate limiters (`signin_ip` 10/min, `signin_email` 5/min, `refresh_token`, `twofa_verification_ip`/`twofa_verification_user`, `password_reset_confirm`, `signout`, `signout_all`) were selected solely by matching the **REST URL path** in `ApiRateLimitRequestResolver` / `ApiRateLimitAuthTargetResolver`. API Platform also exposes the same operations as GraphQL mutations (`signIn`, `refreshToken`, `completeTwoFactor`, `confirmPasswordReset`, `signOut`, `signOutAll`) served at `POST /api/graphql` (PUBLIC_ACCESS). That path matched none of the REST checks, so only the loose `global_api_anonymous` limiter (~100 req/min/IP) applied. An attacker could bypass every tight auth throttle by issuing the equivalent GraphQL mutation — enabling credential stuffing against `signIn`, 2FA brute force against `completeTwoFactor`, refresh-token abuse, and reset-token guessing against `confirmPasswordReset`.

### Fix
- New `App\Shared\Application\Resolver\RateLimit\ApiRateLimitGraphQlResolver` (Application layer, existing `Resolver/RateLimit/` directory). For `POST /api/graphql` it decodes the JSON body, detects sensitive mutation field names in the `query`, and emits the **exact same limiter names and keys** as the REST resolvers:
  - IP-keyed limiter is always applied for a recognised sensitive mutation (hard security floor).
  - Identity keys added when present: lowercased `email` from `variables.input` (sign-in), pending-2FA-derived user id (complete 2FA), JWT subject (sign-out).
- Wired into `ApiRateLimitRequestResolver::resolveEndpointLimiters`, so the existing `ApiRateLimitListener` enforces it with no listener change. REST and GraphQL now share one throttle budget per identity/IP.
- No limiter config, env, REST route, or GraphQL schema changes. Batch GraphQL requests remain rejected by the existing `GraphQLBatchRejectListener` (defense in depth).

### Local verification (one-off containers)
- `phpunit --testsuite=Unit --filter RateLimit`: **OK (155 tests)** — includes 14 new `ApiRateLimitGraphQlResolverTest` cases (positive/negative/edge) + 2 wiring tests.
- `phpunit --testsuite=Unit` (full): **OK (2274 tests)**.
- `deptrac analyse`: **Violations 0**.
- `psalm` (changed files): **No errors found**.
- `php-cs-fixer --dry-run` (changed files): **0 of 6 files**.

### BMAD spec
`specs/security-315-graphql-bypasses-rate-limiters/` — `prd.md` (Problem, FR-1..FR-6, NFR-Security/Compatibility/Maintainability, Out-of-scope) and `stories.md` (per-change stories with positive/negative/edge test mapping and verification commands).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Applied per-endpoint auth rate limiters to GraphQL mutations at POST `/api/graphql` to close a bypass for sign-in, 2FA, refresh-token, password-reset, and sign-out throttles. REST and GraphQL now share the same limiter budgets per IP and identity; each mutation occurrence consumes its own token. Fixes #315.

- **Bug Fixes**
  - Added ApiRateLimitGraphQlResolver to parse the GraphQL body, detect sensitive mutations via a field-capture regex and allowlist, and emit the same limiter names/keys as REST (IP floor; plus `email`, 2FA user, or JWT subject when present). Duplicate occurrences are each throttled.
  - Wired into ApiRateLimitRequestResolver::resolveEndpointLimiters so the existing listener enforces these without changes; multiple sensitive fields in one query are all throttled.
  - No changes to limiter config, REST routes, or GraphQL schema; batch GraphQL requests continue to be rejected.

- **Refactors**
  - Tightened types and guarded array access; removed redundant JSON options. Split tests into behavior and edge-case classes, hoisted the GraphQL request helper and endpoint constant into RateLimitClientTestCase. Addressed final CI quality gates.

<sup>Written for commit 334664dff2271254d4f33b7fbeaddac036e6e413. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/VilnaCRM-Org/user-service/pull/329?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/Shared/Application/Resolver/RateLimit/ApiRateLimitGraphQlResolver.php [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+{
+    private const GRAPHQL_PATH = '/api/graphql';
+
+    /**
+     * Captures every GraphQL field invocation (an identifier directly followed by
+     * an argument list). The allowlist of throttled mutations is enforced by the
+     * match expression in {@see resolveMutationTargets()} so that unknown fields
+     * resolve to no rate-limiter targets.
+     */
+    private const MUTATION_FIELD_PATTERN = '/([A-Za-z_][A-Za-z0-9_]*)\s*\(/';
```
[the remark this answers is not in the document]
Comment (dmytrocraft) [reply src/Shared/Application/Resolver/RateLimit/ApiRateLimitGraphQlResolver.php:32]: Not exploitable: API Platform generates these mutations as `signOutUser(input: signOutUserInput!)` / `signOutAllUser(input: signOutAllUserInput!)` — the `input` argument is non-null with no default, so `webonyx/graphql-php` `ProvidedRequiredArguments` validation rejects any argument-less call (`mutation { signOutUser }`) before it reaches the resolver. A throttled mutation therefore always carries `(input: ...)`, so the `\(` in `MUTATION_FIELD_PATTERN` is always satisfied. No no-argument throttle bypass exists.

File: src/Shared/Application/Resolver/RateLimit/ApiRateLimitGraphQlResolver.php [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        Request $request,
+        string $query,
+        array $input
+    ): array {
+        if (preg_match_all(self::MUTATION_FIELD_PATTERN, $query, $matches) === false) {
+            return [];
+        }
+
+        $targets = [];
+        foreach (array_unique($matches[1]) as $mutation) {
```
[the remark this answers is not in the document]
Comment (dmytrocraft) [reply src/Shared/Application/Resolver/RateLimit/ApiRateLimitGraphQlResolver.php]: Fixed in 1a0482ff. `extractMutationNames()` no longer applies `array_unique`, so each aliased duplicate occurrence of a throttled mutation now yields its own rate-limiter target and consumes a separate token. Aliased duplicates (`a: refreshToken(...) b: refreshToken(...)`) execute server-side N times and now cost N limiter hits. The same-name/different-args form is invalid GraphQL (field-merge conflict) but is still charged per occurrence at the listener boundary, which is the safe direction.

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Review by Kravalg — APPROVED** [review]
